channeldb/payment_control.go

```diff
@@ -637,19 +641,36 @@ func fetchPaymentBucketUpdate(tx kvdb.RwTx, paymentHash lntypes.Hash) (
 
 // nextPaymentSequence returns the next sequence number to store for a new
 // payment.
-func nextPaymentSequence(tx kvdb.RwTx) ([]byte, error) {
-	payments, err := tx.CreateTopLevelBucket(paymentsRootBucket)
-	if err != nil {
-		return nil, err
-	}
+func (p *PaymentControl) nextPaymentSequence() ([]byte, error) {
+	p.paymentSeqMx.Lock()
+	defer p.paymentSeqMx.Unlock()
+
+	// Set a new upper bound in the DB every 1000 payments to avoid
+	// conflicts on the sequence when using etcd.
+	if p.currPaymentSeq == p.storedPaymentSeq {
```

Comment:
Cool, I see how this should speed things up considerably! Worst case here is that we just "waste" 1k sequences if we restart `lnd` directly after setting a new upper bound. But that probably shouldn't matter too much if we have gaps in the payment sequences.

Reply:
Yep, wasting an interval is not a huge issue, since we only require these ids to be monotonic. With etcd there was a huge contention on this sequence key causing all payment transactions queueing up, and retried in the STM. This was one of our major bottlenecks.

Reply:
Yeah this is a nice optimization: allocating things in chunks. Only things that comes to mind here as an edge case is ensuring some migration or deletion logic doesn't depend on them being exactly off by one each time.